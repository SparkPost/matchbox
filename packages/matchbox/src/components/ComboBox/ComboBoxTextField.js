import React from 'react';
import PropTypes from 'prop-types';
import { Label } from '../Label';
import { Error } from '../Error';
import { Tag } from '../Tag';
import { identity, noop } from '../../helpers/event';
import { onKey } from '../../helpers/keyEvents';
import classnames from 'classnames';
import styles from './ComboBoxTextField.module.scss';

function ComboBoxTextField(props) {
  const {
    autoFocus,
    disabled,
    error,
    errorInLabel,
    helpText,
    id,
    itemToString,
    label,
    labelHidden,
    name,
    onChange,
    onFocus,
    onBlur,
    onKeyDown,
    placeholder,
    readOnly,
    removeItem,
    style,
    selectedItems,
    value,
    ...rest
  } = props;

  const inputRef = React.useRef();

  const setWrapperClasses = classnames(
    styles.Wrapper,
    error && styles.Error,
    disabled && styles.Disabled
  );

  const labelMarkup = (
    <Label id={id} label={label}>
      {error && errorInLabel && <Error className={styles.InlineError} wrapper='span' error={error} />}
    </Label>
  );

  const helpMarkup = helpText
    ? <div className={styles.HelpText}>{helpText}</div>
    : null;

  const selectedMarkup = selectedItems.length
    ? selectedItems.map((item, i) => (
      <Tag
        key={i}
        onRemove={!disabled
          ? () => removeItem(item)
          : null
        }
      >
        {itemToString(item)}
      </Tag>
    ))
    : null;

  // Auto focuses the input
  function handleClick() {
    inputRef.current.focus();
  }

  // Removes last item with a backspace and and empty input value
  function handleKeyDown(e) {
    if (!value && selectedItems.length) {
      onKey('backspace', () => {
        removeItem(selectedItems[selectedItems.length - 1]);
      })(e);
    }

    if (onKeyDown) {
      onKeyDown(e);
    }
  }

  return (
    <fieldset className={styles.TextField}>
      {label && !labelHidden && labelMarkup}
      <div className={setWrapperClasses} onClick={handleClick}>
        {selectedMarkup}
        <input
          {...rest}
          autoFocus={autoFocus}
          className={styles.Input}
          disabled={disabled}
          id={id}
          name={name}
          onFocus={onFocus}
          onBlur={onBlur}
          onChange={onChange}
          onKeyDown={handleKeyDown}
          placeholder={placeholder}
          readOnly={readOnly}
          ref={inputRef}
          style={style}
          value={value}
        />
      </div>
      {error && !errorInLabel && <Error error={error} />}
      {helpMarkup}
    </fieldset>
  );
}

ComboBoxTextField.propTypes = {
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  errorInLabel: PropTypes.bool,
  helpText: PropTypes.node,
  id: PropTypes.string,
  itemToString: PropTypes.func,
  label: PropTypes.string,
  labelHidden: PropTypes.bool,
  name: PropTypes.string,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  placeholder: PropTypes.string,
  readOnly: PropTypes.bool,
  removeItem: PropTypes.func,
  selectedItems: PropTypes.array
};

ComboBoxTextField.defaultProps = {
  selectedItems: [],
  itemToString: identity,
  removeItem: noop
};

export default ComboBoxTextField;
