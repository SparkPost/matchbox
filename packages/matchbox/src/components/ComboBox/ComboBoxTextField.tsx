import React from 'react';
import styled from 'styled-components';
import { margin, MarginProps } from 'styled-system';
import { pick } from '@styled-system/props';
import { Box } from '../Box';
import { ErrorLabel } from '../Error';
import { HelpText } from '../HelpText';
import { Inline } from '../Inline';
import { Label } from '../Label';
import { Tag } from '../Tag';
import { identity, noop } from '../../helpers/event';
import { onKey } from '../../helpers/keyEvents';
import { omit } from '../../helpers/props';
import useInputDescribedBy from '../../hooks/useInputDescribedBy';
import { StyledInput, StyledInputWrapper } from './styles';

const StyledWrapper = styled('div')`
  ${margin}
`;

export type ComboBoxTextFieldProps = {
  delimiter?: string;
  error?: string;
  errorInLabel?: boolean;
  helpText?: React.ReactNode;
  id: string;
  itemToString?: (item: any) => string;
  label?: string;
  labelHidden?: boolean;
  removeItem?: (a: React.ReactNode) => void;
  selectedItems?: React.ReactNodeArray;
} & MarginProps &
  React.ComponentPropsWithRef<'input'>;

function ComboBoxTextField(props: ComboBoxTextFieldProps): JSX.Element {
  const {
    autoFocus,
    disabled,
    delimiter,
    children,
    error,
    errorInLabel,
    helpText,
    id,
    itemToString = identity,
    label,
    labelHidden,
    name,
    onChange,
    onFocus,
    onBlur,
    onKeyDown,
    placeholder,
    readOnly,
    removeItem = noop,
    required,
    style,
    selectedItems = [],
    value,
    ...rest
  } = props;

  const inputProps = omit(rest, margin.propNames);
  const systemProps = pick(rest);

  const { describedBy, errorId, helpTextId } = useInputDescribedBy({
    id,
    hasHelpText: !!helpText,
    hasError: !!error,
  });

  // Removes last item with a backspace and and empty input value
  function handleKeyDown(e) {
    if (!value && selectedItems.length) {
      onKey('backspace', () => {
        removeItem(selectedItems[selectedItems.length - 1]);
      })(e);
    }

    if (onKeyDown) {
      onKeyDown(e);
    }
  }

  return (
    <StyledWrapper {...systemProps}>
      {label && (
        <Label id={id} labelHidden={labelHidden}>
          <Box as="span" pr="200">
            {label}
          </Box>
          {required && (
            <Box as="span" pr="200" aria-hidden="true">
              *
            </Box>
          )}
          {error && errorInLabel && (
            <ErrorLabel id={errorId} wrapper="span">
              {error}
            </ErrorLabel>
          )}
        </Label>
      )}
      <StyledInputWrapper $hasError={!!error} $isDisabled={disabled}>
        {selectedItems.length > 0 && (
          <Box display="flex" pl="200" pt="0.375rem">
            <Inline space="100">
              {selectedItems.map((item, i) => {
                return (
                  <span key={i}>
                    <Tag onRemove={!disabled ? () => removeItem(item) : null}>
                      {itemToString(item)}
                    </Tag>
                    {delimiter && i < selectedItems.length - 1 ? (
                      <Box
                        as="span"
                        ml="100"
                        fontWeight="medium"
                        fontStyle="italic"
                        style={{ textTransform: 'uppercase' }}
                      >
                        {delimiter}
                      </Box>
                    ) : null}
                  </span>
                );
              })}
            </Inline>
          </Box>
        )}
        <StyledInput
          {...describedBy}
          autoFocus={autoFocus}
          disabled={disabled}
          id={id}
          name={name}
          onFocus={onFocus}
          onBlur={onBlur}
          onChange={onChange}
          onKeyDown={handleKeyDown}
          placeholder={placeholder}
          readOnly={readOnly}
          required={required}
          style={style}
          type="text"
          value={value}
          {...inputProps}
        />
      </StyledInputWrapper>
      {/* Menu is rendered here so it is positioned correctly before error and helptext */}
      {children}
      {helpText && <HelpText id={helpTextId}>{helpText}</HelpText>}
      {error && !errorInLabel && <ErrorLabel id={errorId}>{error}</ErrorLabel>}
    </StyledWrapper>
  );
}

export default ComboBoxTextField;
