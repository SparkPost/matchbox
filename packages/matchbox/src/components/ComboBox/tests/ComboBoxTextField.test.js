import React from 'react';
import ComboBoxTextField from '../ComboBoxTextField';
import { shallow } from 'enzyme';

describe('ComboBoxTextField', () => {
  let focus;

  beforeEach(() => {
    focus = jest.fn();
    React.useRef = jest.fn(() => ({
      current: {
        focus: focus
      }
    }));
  });

  const items = [
    { content: 'foo' },
    { content: 'bar' }
  ];
  const itemToString = ({ content }) => content;
  const subject = (props = {}) => shallow(
    <ComboBoxTextField
      selectedItems={items}
      itemToString={itemToString}
      {...props}
    />
  );

  it('should render the full fieldset with label, id, error and help text', () => {
    const textfield = subject({
      label: 'test label',
      id: 'test-id',
      helpText: 'help',
      error: 'error'
    });

    expect(textfield).toMatchSnapshot();
  });

  it('should render an error in label', () => {
    const textfield = subject({ error: 'test error', label: 'test label', errorInLabel: true });
    expect(textfield.find('Error').prop('className')).toMatch('InlineError');
  });

  it('should render a hidden label', () => {
    const textfield = subject({ label: 'test label', labelHidden: true });
    expect(textfield.find('Label')).not.toExist();
  });

  it('should focus on input when clicking wrapper', () => {
    const textfield = subject({ });
    textfield.find('.Wrapper').simulate('click');
    expect(focus).toHaveBeenCalled();
  });

  describe('selected items', () => {
    it('should render selected items correctly', () => {
      const textfield = subject();
      expect(textfield.find('Tag')).toMatchSnapshot();
    });

    it('should handle remove on a tag correctly', () => {
      const removeItem = jest.fn();
      const textfield = subject({ removeItem });

      textfield.find('Tag').at(0).simulate('remove');
      expect(removeItem).toHaveBeenCalledWith({ content: 'foo' });
    });

    it('should not be removable when disabled', () => {
      const removeItem = jest.fn();
      const textfield = subject({ removeItem, disabled: true });

      textfield.find('Tag').at(0).simulate('remove');
      expect(removeItem).not.toHaveBeenCalled();
    });

    it('should handle remove on a backspace correctly', () => {
      const removeItem = jest.fn();
      const onKeyDown = jest.fn();
      const textfield = subject({ removeItem, value: '', onKeyDown });

      textfield.find('input').simulate('keyDown', { key: 'Backspace', shiftKey: false });
      expect(removeItem).toHaveBeenCalledWith({ content: 'bar' });
      expect(onKeyDown).toHaveBeenCalledWith({ key: 'Backspace', shiftKey: false });
    });

    it('should not remove on a backspace with a value', () => {
      const removeItem = jest.fn();
      const textfield = subject({ removeItem, value: 'test' });

      textfield.find('input').simulate('keyDown', { key: 'Backspace', shiftKey: false });
      expect(removeItem).not.toHaveBeenCalled();
    });
  });
});
