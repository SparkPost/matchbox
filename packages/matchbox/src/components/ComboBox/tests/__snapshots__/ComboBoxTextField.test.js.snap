// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComboBoxTextField selected items should render selected items correctly 1`] = `
Array [
  <Tag
    key="0"
    onRemove={[Function]}
  >
    foo
  </Tag>,
  <Tag
    key="1"
    onRemove={[Function]}
  >
    bar
  </Tag>,
]
`;

exports[`ComboBoxTextField should render the full fieldset with label, id, error and help text 1`] = `
<fieldset
  className="TextField"
>
  <Label
    id="test-id"
    label="test label"
  />
  <div
    className="Wrapper Error"
    onClick={[Function]}
  >
    <Tag
      key="0"
      onRemove={[Function]}
    >
      foo
    </Tag>
    <Tag
      key="1"
      onRemove={[Function]}
    >
      bar
    </Tag>
    <input
      className="Input"
      id="test-id"
      onKeyDown={[Function]}
    />
  </div>
  <Error
    error="error"
  />
  <div
    className="HelpText"
  >
    help
  </div>
</fieldset>
`;
